hnn_core/parallel_backends.py

```diff
@@ -0,0 +1,304 @@
+        net.spikes._times.append(spikedata[0])
+        net.spikes._gids.append(spikedata[1])
+        net.gid_dict = spikedata[2]  # only have one gid_dict
+        net.spikes.update_types(net.gid_dict)
+
+    return dpls
+
+
+class JoblibBackend(object):
+    """The JoblibBackend class.
+
+    Parameters
+    ----------
+    n_jobs : int
+        The number of jobs to start in parallel
+
+    Attributes
+    ----------
+    n_jobs : int
+        The number of jobs to start in parallel
+    """
+    def __init__(self, n_jobs=1):
+        self.type = 'joblib'
+
+        self.n_jobs = n_jobs
+        print("joblib will run over %d jobs" % (self.n_jobs))
+
+    def _parallel_func(self, func):
+        if self.n_jobs != 1:
+            try:
+                from joblib import Parallel, delayed
+            except ImportError:
+                warn('joblib not installed. Cannot run in parallel.')
+                self.n_jobs = 1
+        if self.n_jobs == 1:
+            my_func = func
+            parallel = list
+        else:
+            parallel = Parallel(self.n_jobs)
+            my_func = delayed(func)
+
+        return parallel, my_func
+
+    def __enter__(self):
+        global BACKEND
+
+        self._old_backend = BACKEND
+        BACKEND = self
+
+        return self
+
+    def __exit__(self, type, value, traceback):
+        global BACKEND
+
+        BACKEND = self._old_backend
+
+    def _clone_and_simulate(self, net, trial_idx):
+        # avoid relative lookups after being forked by joblib
+        from hnn_core.neuron import NeuronNetwork, _simulate_single_trial
+
+        if trial_idx != 0:
+            net.params['prng_*'] = trial_idx
+
+        neuron_net = NeuronNetwork(net)
+        dpl = _simulate_single_trial(neuron_net)
+
+        spikedata = neuron_net.get_data_from_neuron()
+
+        return dpl, spikedata
+
+    def simulate(self, net):
+        """Simulate the HNN model
+
+        Parameters
+        ----------
+        net : Network object
+            The Network object specifying how cells are
+            connected.
+
+        Returns
+        -------
+        dpl: list of Dipole
+            The Dipole results from each simulation trial
+        """
+
+        n_trials = net.params['N_trials']
+        dpls = []
+
+        parallel, myfunc = self._parallel_func(self._clone_and_simulate)
+        sim_data = parallel(myfunc(net, idx) for idx in range(n_trials))
+
+        dpls = _gather_trial_data(sim_data, net, n_trials)
+        return dpls
+
+
+class MPIBackend(object):
+    """The MPIBackend class.
+
+    Parameters
+    ----------
+    n_procs : int | None
+        The number of processes processes MPI will use (spread over cores)
+    mpi_cmd : str
+        The name of the mpi launcher executable. Will use 'mpiexec'
+        (openmpi) by default.
+
+    Attributes
+    ----------
+
+    n_jobs : int
+        The number of jobs to start in parallel (NOT SUPPORTED)
+    n_procs : int
```

Comment:
```suggestion
    n_procs : int | None
```

also say what happens when `n_procs` is `None`